contracts/instruments/RibbonThetaVaultYearn.sol

```diff
@@ -0,0 +1,720 @@
+        // solhint-disable-next-line
+        // https://github.com/AlphaFinanceLab/alphahomora/blob/340653c8ac1e9b4f23d5b81e61307bf7d02a26e8/contracts/5/Bank.sol#L104
+        uint256 share =
+            shareSupply == 0 ? amount : amount.mul(shareSupply).div(total);
+
+        require(
+            shareSupply.add(share) >= MINIMUM_SUPPLY,
+            "Insufficient share supply"
+        );
+
+        emit Deposit(msg.sender, amount, share);
+
+        _mint(msg.sender, share);
+    }
+
+    /**
+     * @notice Withdraws ETH from vault using vault shares
+     * @param share is the number of vault shares to be burned
+     */
+    function withdrawETH(uint256 share) external nonReentrant {
+        require(asset == WETH, "!WETH");
+        uint256 withdrawAmount = _withdraw(share, false);
+
+        IWETH(WETH).withdraw(withdrawAmount);
+        (bool success, ) = msg.sender.call{value: withdrawAmount}("");
+        require(success, "ETH transfer failed");
+    }
+
+    /**
+     * @notice Withdraws WETH from vault using vault shares
+     * @param share is the number of vault shares to be burned
+     */
+    function withdraw(uint256 share) external nonReentrant {
+        uint256 withdrawAmount = _withdraw(share, false);
+        IERC20(asset).safeTransfer(msg.sender, withdrawAmount);
+    }
+
+    /**
+     * @notice Burns vault shares and checks if eligible for withdrawal
+     * @param share is the number of vault shares to be burned
+     * @param isScheduled is whether the withdraw was scheduled
+     */
+    function _withdraw(uint256 share, bool isScheduled)
+        private
+        returns (uint256)
+    {
+        (uint256 amountAfterFee, uint256 feeAmount) =
+            withdrawAmountWithShares(share);
+
+        emit Withdraw(msg.sender, amountAfterFee, share, feeAmount);
+
+        _burn(isScheduled ? address(this) : msg.sender, share);
+
+        _unwrapYieldToken(amountAfterFee.add(feeAmount));
+
+        IERC20(asset).safeTransfer(feeRecipient, feeAmount);
+
+        return amountAfterFee;
+    }
+
+    /**
+     * @notice Unwraps the necessary amount of the yield-bearing yearn token
+     *         and transfers amount to relevant recipient
+     * @param amount is the amount of `asset` to withdraw
+     */
+    function _unwrapYieldToken(uint256 amount) private returns (uint256) {
+        uint256 assetBalance = IERC20(asset).balanceOf(address(this));
+        uint256 amountToUnwrap =
+            wdiv(
+                max(assetBalance, amount).sub(assetBalance),
+                collateralToken.pricePerShare()
+            );
+        amountToUnwrap = amountToUnwrap.add(
+            amountToUnwrap.mul(YEARN_WITHDRAWAL_SLIPPAGE).div(10000)
+        );
+        if (amountToUnwrap > 0) {
+            collateralToken.withdraw(
+                amountToUnwrap,
+                address(this),
+                YEARN_WITHDRAWAL_SLIPPAGE
+            );
+        }
+        return amountToUnwrap;
```

Comment:
Question, can we rely on yVaults to return us the exact `amountToUnwrap` tokens to us? We should revert if it returns less in reality:

```solidity
// in _unwrapYieldToken
minUnwrappedAmount = max(assetBalance, amount).sub(assetBalance)
uint beforeBalance = IERC20(asset).balanceOf(address(this))
collateralToken.withdraw(...)
uint afterBalance = IERC20(asset).balanceOf(address(this))
uint change = afterBalance.sub(beforeBalance)

require(change >= minUnwrappedAmount, "not enough")

return change // return the actual balance change instead of the calculated amount
```

Reply:
theres an assert in withdraw() method which reverts if losses more than what we consider acceptable (using slippage param) ([line 1035](https://etherscan.io/address/0xa9fe4601811213c340e850ea305481aff02f5b28#code))